integrations/spark-extensions-base/src/main/scala/org/apache/spark/sql/execution/datasources/v2/NessieUtils.scala

```diff
@@ -234,6 +256,92 @@ object NessieUtils {
     )
   }
 
+  def getBaseIcebergCatalog(
+      currentCatalog: CatalogPlugin,
+      catalog: Option[String]
+  ): Option[Any] = {
+    val catalogName = catalog.getOrElse(currentCatalog.name)
+    val catalogImpl =
+      SparkSession.active.sessionState.catalogManager.catalog(catalogName)
+
+    getBaseIcebergCatalog(catalogImpl)
+  }
+
+  def getBaseIcebergCatalog(
+      catalogImpl: CatalogPlugin
+  ): Option[Any] = {
+    try {
+      // `catalogImpl` is (should be) an org.apache.iceberg.spark.SparkCatalog.
+      // We need the Iceberg `Catalog` instance from `SparkCatalog`
+      var icebergCatalog = icebergCatalogFromSparkCatalog(catalogImpl)
+
+      // If the Iceberg catalog is a `CachingCatalog`, get the "base" catalog from it.
+      if (icebergCatalog.getClass.getSimpleName.equals("CachingCatalog")) {
+        val cachingCatalogCatalog =
+          icebergCatalog.getClass.getDeclaredField("catalog")
+        cachingCatalogCatalog.setAccessible(true)
+        icebergCatalog = cachingCatalogCatalog.get(icebergCatalog)
+      }
+
+      Some(icebergCatalog)
+    } catch {
+      // TODO have something better, at least log something
+      case _: Exception => None
+    }
+  }
+
+  private def icebergCatalogFromSparkCatalog(
+      catalogImpl: CatalogPlugin
+  ): Any = {
+    // `catalogImpl` is an `org.apache.iceberg.spark.SparkCatalog`...
+    try {
+      // ... and most(!!) implementations of `o.a.i.s.SparkCatalog` have a
+      // `public Catalog icebergCatalog()` function...
+      catalogImpl.getClass
```

Comment:
Instead of that, can't you test if the impl implements `HasIcebergCatalog`? Both `SparkCatalog` and `SparkSessionCatalog` do.

Reply:
See all the different source files for all the Spark versions. Iceberg's code base is ... not great.

Reply:
Still no:
```
$ find . -name HasIcebergCatalog.java
./spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/HasIcebergCatalog.java
./spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/HasIcebergCatalog.java
./spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/source/HasIcebergCatalog.java
```

No `HasIcebergCatalog` for 3.1